File: fsharp/accumulate/Accumulate.fs
module Accumulate

let accumulate (func: 'a -> 'b) (input: 'a list): 'b list =
    // https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/lists
    // The use of the accumulator argument makes the function tail recursive.
    let rec loop list acc =
        match list with
        | head :: tail -> loop tail (func head :: acc)
        | [] -> List.rev acc
    loop input []

File: erlang/accumulate/src/accumulate.erl
-module(accumulate).

-export([accumulate/2]).

accumulate(Fn, List) -> accumulate(Fn, List, []).

accumulate(_Fn, [], Accumulator) -> lists:reverse(Accumulator);
accumulate(Fn, [Head|Tail], Accumulator) -> accumulate(Fn, Tail, [Fn(Head) | Accumulator]).

